src/hotspot/share/gc/shenandoah/shenandoahFreeSet.hpp

```diff
@@ -34,24 +34,51 @@ class ShenandoahFreeSet : public CHeapObj<mtGC> {
   ShenandoahHeap* const _heap;
   CHeapBitMap _mutator_free_bitmap;
   CHeapBitMap _collector_free_bitmap;
+  // We keep the _old_collector regions separate from the young collector regions.  This allows us to pack the old regions
+  // further to the right than the young collector regions.  This is desirable because the old collector regions are recycled
+  // even less frequently than the young survivor regions.  In generational mode, the young survivor regions are typically
+  // recycled after tenure age GC passes.
+  CHeapBitMap _old_collector_free_bitmap;
   size_t _max;
 
   // Left-most and right-most region indexes. There are no free regions outside
   // of [left-most; right-most] index intervals
   size_t _mutator_leftmost, _mutator_rightmost;
   size_t _collector_leftmost, _collector_rightmost;
+  size_t _old_collector_leftmost, _old_collector_rightmost;
```

Comment:
Would suggest encapsulation in an interval struct with left & right for each of the three intervals.

At that point, it would make sense to talk about whether these intervals can overlap (because of interspersing of the range of indices in a specific subset), and whether there are invariants/properties of the union of the intervals that can be checked as invariants after these sub-intervals are adjusted.

Reply:
I see that some invariants are checked in `assert_bounds` method. A high-level articulation of those invariants before that method would be useful.